[roll blocker] C:\b\s\w\ir\x\t\5f63de18\debug.dll : fatal error LNK1107: invalid or corrupt file: cannot read at 0x300

```
/==================================================================================================================\
| native_assets_builder/test/test_data/native_dynamic_linking_test is new and failed (RuntimeError, expected Pass) |
\==================================================================================================================/

--- Command "vm" (took 10.000881s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set CHROME_PATH=C:\b\s\w\ir\cache\builder\sdk\third_party\browsers\chrome\Chrome\Application\chrome.exe & out\ReleaseX64\dart-sdk\bin\dart.exe --enable_asserts --sound-null-safety -Dtest_runner.configuration=unittest-asserts-release-win-x64 --ignore-unrecognized-flags --packages=C:\b\s\w\ir\cache\builder\sdk\.dart_tool\package_config.json C:\b\s\w\ir\cache\builder\sdk\third_party\pkg\native\pkgs\native_assets_builder\test\test_data\native_dynamic_linking_test.dart

exit code:
255

stdout:
00:00 [32m+0[0m: native_dynamic_linking build dry_run[0m

00:02 [32m+1[0m: native_dynamic_linking build[0m

Using compiler C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe from BuildConfig.cCompiler.cc.

Trying to recognize file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe.

Tool instance file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe is likely Tool(cl.exe).

Looking up version with --version for ToolInstance(cl.exe, null, file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe).

Running `C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe`.

Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64

Copyright (C) Microsoft Corporation.  All rights reserved.



usage: cl [ option... ] filename... [ /link linkoption... ]

Found version for ToolInstance(cl.exe, 19.34.31937, file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe).

Using envScript from config: file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/Windows%20Kits/10/bin/SetEnv.cmd

Using envScriptArgs from config: [/x64, /x64]

Running `(cd C:\b\s\w\ir\x\t\5f63de18\; INCLUDE=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\um;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\shared;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\winrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\ucrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\include;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\include LIB=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\um\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\ucrt\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\lib\x64 LIBPATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x86\store\references;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\UnionMetadata\10.0.22621.0 PATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\bin\10.0.22621.0\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64;C:\b\s\w\ir\bbagent_utility_packages;C:\b\s\w\ir\bbagent_utility_packages\bin;C:\b\s\w\ir\cipd_bin_packages;C:\b\s\w\ir\cipd_bin_packages\bin;C:\b\s\w\ir\cipd_bin_packages\cpython3;C:\b\s\w\ir\cipd_bin_packages\cpython3\bin;C:\b\s\w\ir\cache\cipd_client;C:\b\s\w\ir\cache\cipd_client\bin;C:\b\s\cipd_cache\bin;C:\Program Files\OpenSSH\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Puppet Labs\Puppet\bin;C:\b\s\w\ir\kitchen-checkout\depot_tools VCINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\ VCToolsInstallDir=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\ VSINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\.\ C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe /DDEBUG C:\b\s\w\ir\cache\builder\sdk\third_party\pkg\native\pkgs\native_assets_builder\test_data\native_dynamic_linking\src\debug.c /link /DLL /out:C:\b\s\w\ir\x\t\5f63de18\debug.dll )`.

Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64

Copyright (C) Microsoft Corporation.  All rights reserved.



debug.c

Microsoft (R) Incremental Linker Version 14.34.31937.0

Copyright (C) Microsoft Corporation.  All rights reserved.



/out:debug.exe 

/DLL 

/out:C:\b\s\w\ir\x\t\5f63de18\debug.dll 

debug.obj 

   Creating library C:\b\s\w\ir\x\t\5f63de18\debug.lib and object C:\b\s\w\ir\x\t\5f63de18\debug.exp

Using compiler C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe from BuildConfig.cCompiler.cc.

Trying to recognize file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe.

Tool instance file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe is likely Tool(cl.exe).

Looking up version with --version for ToolInstance(cl.exe, null, file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe).

Running `C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe`.

Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64

Copyright (C) Microsoft Corporation.  All rights reserved.



usage: cl [ option... ] filename... [ /link linkoption... ]

Found version for ToolInstance(cl.exe, 19.34.31937, file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/VC/Tools/MSVC/14.34.31933/bin/Hostx64/x64/cl.exe).

Using envScript from config: file:///C:/b/s/w/ir/cache/builder/sdk/win_toolchain/vs_files/27370823e7/Windows%20Kits/10/bin/SetEnv.cmd

Using envScriptArgs from config: [/x64, /x64]

Running `(cd C:\b\s\w\ir\x\t\5f63de18\; INCLUDE=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\um;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\shared;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\winrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\ucrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\include;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\include LIB=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\um\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\ucrt\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\lib\x64 LIBPATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x86\store\references;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\UnionMetadata\10.0.22621.0 PATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\bin\10.0.22621.0\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64;C:\b\s\w\ir\bbagent_utility_packages;C:\b\s\w\ir\bbagent_utility_packages\bin;C:\b\s\w\ir\cipd_bin_packages;C:\b\s\w\ir\cipd_bin_packages\bin;C:\b\s\w\ir\cipd_bin_packages\cpython3;C:\b\s\w\ir\cipd_bin_packages\cpython3\bin;C:\b\s\w\ir\cache\cipd_client;C:\b\s\w\ir\cache\cipd_client\bin;C:\b\s\cipd_cache\bin;C:\Program Files\OpenSSH\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Puppet Labs\Puppet\bin;C:\b\s\w\ir\kitchen-checkout\depot_tools VCINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\ VCToolsInstallDir=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\ VSINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\.\ C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe C:\b\s\w\ir\x\t\5f63de18\debug.dll /DDEBUG C:\b\s\w\ir\cache\builder\sdk\third_party\pkg\native\pkgs\native_assets_builder\test_data\native_dynamic_linking\src\math.c /link /DLL /out:C:\b\s\w\ir\x\t\5f63de18\math.dll )`.

Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31937 for x64

Copyright (C) Microsoft Corporation.  All rights reserved.



cl : Command line warning D9024 : unrecognized source file type 'C:\b\s\w\ir\x\t\5f63de18\debug.dll', object file assumed

math.c

Microsoft (R) Incremental Linker Version 14.34.31937.0

Copyright (C) Microsoft Corporation.  All rights reserved.



/out:debug.exe 

/DLL 

/out:C:\b\s\w\ir\x\t\5f63de18\math.dll 

C:\b\s\w\ir\x\t\5f63de18\debug.dll 

math.obj 

C:\b\s\w\ir\x\t\5f63de18\debug.dll : fatal error LNK1107: invalid or corrupt file: cannot read at 0x300



Unhandled exception:

ProcessException: Full command string: '(cd C:\b\s\w\ir\x\t\5f63de18\; INCLUDE=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\um;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\shared;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\winrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Include\10.0.22621.0\ucrt;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\include;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\include LIB=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\um\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\Lib\10.0.22621.0\ucrt\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\atlmfc\lib\x64 LIBPATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\lib\x86\store\references;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\UnionMetadata\10.0.22621.0 PATH=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\Windows Kits\10\bin\10.0.22621.0\x64;C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\HostX64\x64;C:\b\s\w\ir\bbagent_utility_packages;C:\b\s\w\ir\bbagent_utility_packages\bin;C:\b\s\w\ir\cipd_bin_packages;C:\b\s\w\ir\cipd_bin_packages\bin;C:\b\s\w\ir\cipd_bin_packages\cpython3;C:\b\s\w\ir\cipd_bin_packages\cpython3\bin;C:\b\s\w\ir\cache\cipd_client;C:\b\s\w\ir\cache\cipd_client\bin;C:\b\s\cipd_cache\bin;C:\Program Files\OpenSSH\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Puppet Labs\Puppet\bin;C:\b\s\w\ir\kitchen-checkout\depot_tools VCINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\ VCToolsInstallDir=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\ VSINSTALLDIR=C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\.\ C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe C:\b\s\w\ir\x\t\5f63de18\debug.dll /DDEBUG C:\b\s\w\ir\cache\builder\sdk\third_party\pkg\native\pkgs\native_assets_builder\test_data\native_dynamic_linking\src\math.c /link /DLL /out:C:\b\s\w\ir\x\t\5f63de18\math.dll )'.

Exit code: '2'.

For the output of the process check the logger output.

  Command: C:\b\s\w\ir\cache\builder\sdk\win_toolchain\vs_files\27370823e7\VC\Tools\MSVC\14.34.31933\bin\Hostx64\x64\cl.exe C:\b\s\w\ir\x\t\5f63de18\debug.dll /DDEBUG C:\b\s\w\ir\cache\builder\sdk\third_party\pkg\native\pkgs\native_assets_builder\test_data\native_dynamic_linking\src\math.c /link /DLL /out:C:\b\s\w\ir\x\t\5f63de18\math.dll

#0      runProcess (package:native_toolchain_c/src/utils/run_process.dart:94:5)

<asynchronous suspension>

dart-lang/sdk#1      RunCBuilder.runCl (package:native_toolchain_c/src/cbuilder/run_cbuilder.dart:327:20)

<asynchronous suspension>

dart-lang/sdk#2      RunCBuilder.run (package:native_toolchain_c/src/cbuilder/run_cbuilder.dart:121:7)

<asynchronous suspension>

dart-lang/sdk#3      CBuilder.run (package:native_toolchain_c/src/cbuilder/cbuilder.dart:161:7)

<asynchronous suspension>

dart-lang/sdk#4      main.<anonymous closure> (file:///C:/b/s/w/ir/cache/builder/sdk/third_party/pkg/native/pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart:48:7)

<asynchronous suspension>

dart-lang/sdk#5      build (package:native_assets_cli/src/api/build.dart:93:3)

<asynchronous suspension>

dart-lang/sdk#6      main (file:///C:/b/s/w/ir/cache/builder/sdk/third_party/pkg/native/pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart:10:3)

<asynchronous suspension>



255

00:07 [32m+1[0m[31m -1[0m: native_dynamic_linking build [1m[31m[E][0m[0m

  Expected: <0>
    Actual: <255>
  

  package:matcher/src/expect/expect.dart 149:31                                                           fail
  package:matcher/src/expect/expect.dart 144:3                                                            _expect
  package:matcher/src/expect/expect.dart 56:3                                                             expect
  third_party\pkg\native\pkgs\native_assets_builder\test\test_data\native_dynamic_linking_test.dart 76:7  main.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/declarer.dart 229:9                                                        Declarer.test.<fn>.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/declarer.dart 227:7                                                        Declarer.test.<fn>
  ===== asynchronous gap ===========================
  package:test_api/src/backend/invoker.dart 258:9                                                         Invoker._waitForOutstandingCallbacks.<fn>
  

00:07 [32m+1[0m[31m -1[0m: [31mSome tests failed.[0m



Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.

stderr:
Unhandled exception:
Dummy exception to set exit code.

--- Re-run this test:
python3 tools/test.py -n unittest-asserts-release-win-x64 native_assets_builder/test/test_data/native_dynamic_linking_test
```

[logs](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8737612099528276721/+/u/test_results/new_test_failures__logs_)

--- a/pkgs/native_assets_builder/dart_test.yaml
+++ b/pkgs/native_assets_builder/dart_test.yaml
@@ -1,3 +1,4 @@
 paths:
   - test/build_runner/
   - test/model/
+  - test/test_data/

--- a/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_helper.dart
+++ b/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_helper.dart
@@ -11,5 +11,5 @@ void main(List<String> arguments) {
   final addLibrary = DynamicLibrary.open(addLibraryPath);
   final add = addLibrary.lookupFunction<Int32 Function(Int32, Int32),
       int Function(int, int)>('add');
-  print(add(a, b));
+  print('Result: ${add(a, b)}');
 }

--- a/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart
+++ b/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart
@@ -17,110 +17,111 @@ import 'package:test/test.dart';
 import '../helpers.dart';
 
 void main() async {
-  late Uri tempUri;
   const name = 'native_dynamic_linking';
-
-  setUp(() async {
-    tempUri = (await Directory.systemTemp.createTemp()).uri;
-  });
-
-  tearDown(() async {
-    await Directory.fromUri(tempUri).delete(recursive: true);
-  });
 
   for (final dryRun in [true, false]) {
     final testSuffix = dryRun ? ' dry_run' : '';
-    test('native_dynamic_linking build$testSuffix', () async {
-      final testTempUri = tempUri.resolve('test1/');
-      await Directory.fromUri(testTempUri).create();
-      final testPackageUri = testDataUri.resolve('$name/');
-      final dartUri = Uri.file(Platform.resolvedExecutable);
+    test(
+      'native_dynamic_linking build$testSuffix',
+      () => inTempDir((tempUri) async {
+        final testTempUri = tempUri.resolve('test1/');
+        await Directory.fromUri(testTempUri).create();
+        final testPackageUri = testDataUri.resolve('$name/');
+        final dartUri = Uri.file(Platform.resolvedExecutable);
 
-      final config = BuildConfigImpl(
-        outputDirectory: tempUri,
-        packageName: name,
-        packageRoot: testPackageUri,
-        targetOS: OSImpl.current,
-        version: HookConfigImpl.latestVersion,
-        linkModePreference: LinkModePreferenceImpl.dynamic,
-        dryRun: dryRun,
-        linkingEnabled: false,
-        targetArchitecture: dryRun ? null : ArchitectureImpl.current,
-        buildMode: dryRun ? null : BuildModeImpl.debug,
-        cCompiler: dryRun
-            ? null
-            : CCompilerConfigImpl(
-                compiler: cc,
-                envScript: envScript,
-                envScriptArgs: envScriptArgs,
-              ),
-      );
+        final config = BuildConfigImpl(
+          outputDirectory: tempUri,
+          packageName: name,
+          packageRoot: testPackageUri,
+          targetOS: OSImpl.current,
+          version: HookConfigImpl.latestVersion,
+          linkModePreference: LinkModePreferenceImpl.dynamic,
+          dryRun: dryRun,
+          linkingEnabled: false,
+          targetArchitecture: dryRun ? null : ArchitectureImpl.current,
+          buildMode: dryRun ? null : BuildModeImpl.debug,
+          cCompiler: dryRun
+              ? null
+              : CCompilerConfigImpl(
+                  compiler: cc,
+                  envScript: envScript,
+                  envScriptArgs: envScriptArgs,
+                ),
+        );
 
-      final buildConfigUri = testTempUri.resolve('build_config.json');
-      File.fromUri(buildConfigUri)
-          .writeAsStringSync(jsonEncode(config.toJson()));
+        final buildConfigUri = testTempUri.resolve('build_config.json');
+        File.fromUri(buildConfigUri)
+            .writeAsStringSync(jsonEncode(config.toJson()));
 
-      final processResult = await Process.run(
-        dartUri.toFilePath(),
-        [
-          'hook/build.dart',
-          '--config=${buildConfigUri.toFilePath()}',
-        ],
-        workingDirectory: testPackageUri.toFilePath(),
-      );
-      if (processResult.exitCode != 0) {
-        print(processResult.stdout);
-        print(processResult.stderr);
-        print(processResult.exitCode);
-      }
-      expect(processResult.exitCode, 0);
-
-      final buildOutputUri = tempUri.resolve('build_output.json');
-      final buildOutput = HookOutputImpl.fromJsonString(
-          await File.fromUri(buildOutputUri).readAsString());
-      final assets = buildOutput.assets;
-      final dependencies = buildOutput.dependencies;
-      if (dryRun) {
-        expect(assets.length, greaterThanOrEqualTo(3));
-        expect(dependencies, <Uri>[]);
-      } else {
-        expect(assets.length, 3);
-        expect(await assets.allExist(), true);
-        expect(
-          dependencies,
+        final processResult = await Process.run(
+          dartUri.toFilePath(),
           [
-            testPackageUri.resolve('src/debug.c'),
-            testPackageUri.resolve('src/math.c'),
-            testPackageUri.resolve('src/add.c'),
+            'hook/build.dart',
+            '--config=${buildConfigUri.toFilePath()}',
           ],
+          workingDirectory: testPackageUri.toFilePath(),
         );
+        if (processResult.exitCode != 0) {
+          print(processResult.stdout);
+          print(processResult.stderr);
+          print(processResult.exitCode);
+        }
+        expect(processResult.exitCode, 0);
 
-        final addLibraryPath = assets
-            .firstWhere((asset) => asset.id.endsWith('add.dart'))
-            .file!
-            .toFilePath();
-        final addResult = await runProcess(
-          executable: dartExecutable,
-          arguments: [
-            'run',
-            pkgNativeAssetsBuilderUri
-                .resolve('test/test_data/native_dynamic_linking_add.dart')
-                .toFilePath(),
-            addLibraryPath,
-            '1',
-            '2',
-          ],
-          environment: {
-            // Add the directory containing the linked dynamic libraries to the
-            // PATH so that the dynamic linker can find them.
-            if (Platform.isWindows)
-              'PATH': '${tempUri.toFilePath()};${Platform.environment['PATH']}',
-          },
-          throwOnUnexpectedExitCode: true,
-          logger: logger,
-        );
-        expect(addResult.stdout, 'Adding 1 and 2.\n3\n');
-      }
-    });
+        final buildOutputUri = tempUri.resolve('build_output.json');
+        final buildOutput = HookOutputImpl.fromJsonString(
+            await File.fromUri(buildOutputUri).readAsString());
+        final assets = buildOutput.assets;
+        final dependencies = buildOutput.dependencies;
+        if (dryRun) {
+          expect(assets.length, greaterThanOrEqualTo(3));
+          expect(dependencies, <Uri>[]);
+        } else {
+          expect(assets.length, 3);
+          expect(await assets.allExist(), true);
+          expect(
+            dependencies,
+            [
+              testPackageUri.resolve('src/debug.c'),
+              testPackageUri.resolve('src/math.c'),
+              testPackageUri.resolve('src/add.c'),
+            ],
+          );
+
+          final addLibraryPath = assets
+              .firstWhere((asset) => asset.id.endsWith('add.dart'))
+              .file!
+              .toFilePath();
+          final addResult = await runProcess(
+            executable: dartExecutable,
+            arguments: [
+              'run',
+              pkgNativeAssetsBuilderUri
+                  .resolve('test/test_data/native_dynamic_linking_helper.dart')
+                  .toFilePath(),
+              addLibraryPath,
+              '1',
+              '2',
+            ],
+            environment: {
+              // Add the directory containing the linked dynamic libraries to the
+              // PATH so that the dynamic linker can find them.
+              if (Platform.isWindows)
+                'PATH':
+                    '${tempUri.toFilePath()};${Platform.environment['PATH']}',
+            },
+            throwOnUnexpectedExitCode: true,
+            logger: logger,
+          );
+          expect(
+            addResult.stdout,
+            allOf(
+              contains('Adding 1 and 2.'),
+              contains('Result: 3'),
+            ),
+          );
+        }
+      }),
+    );
   }
 }

--- a/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart
+++ b/pkgs/native_assets_builder/test/test_data/native_dynamic_linking_test.dart
@@ -104,8 +104,8 @@ void main() async {
               '2',
             ],
             environment: {
-              // Add the directory containing the linked dynamic libraries to the
-              // PATH so that the dynamic linker can find them.
+              // Add the directory containing the linked dynamic libraries to
+              // the PATH so that the dynamic linker can find them.
               if (Platform.isWindows)
                 'PATH':
                     '${tempUri.toFilePath()};${Platform.environment['PATH']}',

--- a/pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart
+++ b/pkgs/native_assets_builder/test_data/native_dynamic_linking/hook/build.dart
@@ -66,7 +66,7 @@ extension on BuildConfig {
             '-l$libraryName',
           ],
         OS.windows => [
-            outputDirectory.resolve('$libraryName.dll').toFilePath(),
+            outputDirectory.resolve('$libraryName.lib').toFilePath(),
           ],
         _ => throw UnimplementedError('Unsupported OS: $targetOS'),
       };

